gbpservice/nfp/configurator/drivers/loadbalancer/v2/haproxy/haproxy_driver.py

```diff
@@ -0,0 +1,594 @@
+from gbpservice.nfp.configurator.drivers.loadbalancer.v2.haproxy.\
+    rest_api_driver import HaproxyAmphoraLoadBalancerDriver
+from gbpservice.nfp.configurator.lib import lbv2_constants
+
+DRIVER_NAME = 'loadbalancerv2'
+
+LOG = logging.getLogger(__name__)
+
+
+# As we use the rest client and amphora image from Octavia,
+# we need to have a helper class to simulate Octavia DB operation
+# in order to get Octavia data models from Neutron-lbaas data models
+class OctaviaDataModelBuilder(object):
+
+    def __init__(self, driver=None):
+        self.driver = driver
+
+    # All Octavia data models have these attributes
+    def _get_common_args(self, obj):
+        return {
+            'id': obj.id,
+            'project_id': obj.tenant_id,
+            'name': obj.name,
+            'description': obj.description,
+            'enabled': obj.admin_state_up,
+            'operating_status': obj.operating_status,
+        }
+
+    # Update Octavia model from dict
+    def _update(self, octavia_data_model, update_dict):
+        for key, value in update_dict.items():
+            setattr(octavia_data_model, key, value)
+        return octavia_data_model
+
+    # Translate loadbalancer neutron model dict to octavia model
+    def get_loadbalancer_octavia_model(self, loadbalancer_dict):
+        loadbalancer = n_data_models.LoadBalancer.from_dict(
+            copy.deepcopy(loadbalancer_dict))
+        ret = o_data_models.LoadBalancer()
+        args = self._get_common_args(loadbalancer)
+        vip = o_data_models.Vip(
+            load_balancer_id=loadbalancer.id,
+            ip_address=loadbalancer.vip_address,
+            subnet_id=loadbalancer.vip_subnet_id,
+            port_id=loadbalancer.vip_port.id,
+            load_balancer=ret
+        )
+        amphorae = self.driver.get_amphora(loadbalancer.id)
+        if not amphorae:
+            raise exceptions.IncompleteData(
+                "Amphora information is missing")
+        # TODO: vrrp_group, topology, server_group_id are not included yet
+        args.update({
+            'vip': vip,
+            'amphorae': amphorae,
+            'provisioning_status': loadbalancer.provisioning_status,
+        })
+        if loadbalancer_dict.get('listeners'):
+            listeners = []
+            pools = []
+            for listener_dict in loadbalancer_dict.get('listeners'):
+                listener = self.get_listener_octavia_model(listener_dict)
+                listener.load_balancer = ret
+                listeners.append(listener)
+                pools.extend(listener.pools)
+                for pool in listener.pools:
+                    if pool.id not in [pool.id for pool in pools]:
+                        pools.append(pool)
+            args.update({
+                'listeners': listeners,
+                'pools': pools,
+            })
+
+        ret = self._update(ret, args)
+        return ret
+
+    # Translate listener neutron model dict to octavia model
```

Comment:
This haproxy_driver calls Octavia's rest_api_driver.py to talk to amphora(haproxy vm). rest_api_driver.py can only understand the Octavia data model, as in Octavia the driver gets required information from Octavia's db. That's the reason we need to translate neutron model to Octavia model.